src/common/url-base64-functions.ts

```diff
@@ -0,0 +1,168 @@
+    : _TE
+    ? (s: string) => _fromUint8Array(_TE.encode(s))
+    : (s: string) => _btoa(utob(s))
+
+const re_btou = /[\xC0-\xDF][\x80-\xBF]|[\xE0-\xEF][\x80-\xBF]{2}|[\xF0-\xF7][\x80-\xBF]{3}/g
+const cb_btou = (cccc: string) => {
+    switch (cccc.length) {
+        case 4:
+            const cp =
+                ((0x07 & cccc.charCodeAt(0)) << 18) |
+                ((0x3f & cccc.charCodeAt(1)) << 12) |
+                ((0x3f & cccc.charCodeAt(2)) << 6) |
+                (0x3f & cccc.charCodeAt(3))
+            const offset = cp - 0x10000
+            return _fromCC((offset >>> 10) + 0xd800) + _fromCC((offset & 0x3ff) + 0xdc00)
+        case 3:
+            return _fromCC(
+                ((0x0f & cccc.charCodeAt(0)) << 12) |
+                    ((0x3f & cccc.charCodeAt(1)) << 6) |
+                    (0x3f & cccc.charCodeAt(2)),
+            )
+        default:
+            return _fromCC(((0x1f & cccc.charCodeAt(0)) << 6) | (0x3f & cccc.charCodeAt(1)))
+    }
+}
+const btou = (b: string) => b.replace(re_btou, cb_btou)
+
+const atobPolyfill = (asc: string) => {
+    asc = asc.replace(/\s+/g, '')
+    if (!b64re.test(asc)) throw new TypeError('malformed base64.')
+    asc += '=='.slice(2 - (asc.length & 3))
+    let bin = '',
+        r1,
+        r2,
+        u24
+    for (let i = 0; i < asc.length; ) {
+        u24 =
+            (b64tab[asc.charAt(i++)] << 18) |
+            (b64tab[asc.charAt(i++)] << 12) |
+            ((r1 = b64tab[asc.charAt(i++)]) << 6) |
+            (r2 = b64tab[asc.charAt(i++)])
+        bin +=
+            r1 === 64
+                ? _fromCC((u24 >> 16) & 255)
+                : r2 === 64
+                ? _fromCC((u24 >> 16) & 255, (u24 >> 8) & 255)
+                : _fromCC((u24 >> 16) & 255, (u24 >> 8) & 255, u24 & 255)
+    }
+    return bin
+}
+
+const _atob = _hasatob
+    ? (asc: string) => atob(_tidyB64(asc))
+    : _hasBuffer
+    ? (asc: string) => Buffer.from(asc, 'base64').toString('binary')
+    : atobPolyfill
+
+const _toUint8Array = _hasBuffer
+    ? (a: string) => _U8Afrom(Buffer.from(a, 'base64'))
+    : (a: string) => _U8Afrom(_atob(a), (c) => c?.charCodeAt(0))
+
+const _decode = _hasBuffer
+    ? (a: string) => Buffer.from(a, 'base64').toString('utf8')
+    : _TD
+    ? (a: string) => _TD.decode(_toUint8Array(a))
+    : (a: string) => btou(_atob(a))
+const _unURI = (a: string) => _tidyB64(a?.replace(/[-_]/g, (m0) => (m0 == '-' ? '+' : '/')))
+
+const encode = (src: string, urlsafe = false) => (urlsafe ? _mkUriSafe(_encode(src)) : _encode(src))
+
+const decode = (src: string) => _decode(_unURI(src))
+
+const isValid = (src: any) => {
+    if (typeof src !== 'string') return false
+    const s = src?.replace(/\s+/g, '')?.replace(/={0,2}$/, '')
+    return !/[^\s0-9a-zA-Z\+/]/.test(s) || !/[^\s0-9a-zA-Z\-_]/.test(s)
+}
+export { decode as fromBase64 }
+export { encode as toBase64 }
+export { encode }
+export { decode }
+export { isValid }
```

Comment:
can you please provide short simple `jsDoc` for the functions as well?
it's gonna be hard for people to understand the whole module and what it does in the future. 
also the namings of the functions and variables can be much better, and please add mdn links for `atob` ,`btoa` functions. thanks thanks :pray: